File: haven/storage.py
"""Local, encrypted-at-rest chat history and contact/session state.

Two different encryption layers are at play in this app, deliberately:

  1. The ratchet cipher in crypto.py protects messages ON THE WIRE (and
     gives forward secrecy) — but ratchet keys are one-way hash chains by
     design, so once they've advanced you can no longer re-decrypt an old
     wire ciphertext. That's a feature for transit security, not something
     you want for "let me scroll up and re-read what my friend said".

  2. So on receipt (or send), we decrypt to plaintext once and re-encrypt
     it with a LOCAL storage key (AES-256-GCM, derived from your identity
     key via HKDF — never your raw password) before writing it to disk.
     This is the actual persisted "chat history" / backup material: it's
     encrypted at rest, but independent of the wire ratchet's one-way
     advancement.

Session (ratchet) state is persisted the same way so a restart doesn't
lose the ability to keep chatting with a contact.
"""

from __future__ import annotations

import json
import sqlite3
import threading
import time
from pathlib import Path

from . import crypto


class Store:
    def __init__(self, data_dir: Path, identity: crypto.KeyPair):
        self.storage_key = crypto._hkdf(identity.private_bytes, info=b"local-storage-v1")
        self.db_path = data_dir / "haven.db"
        # Accessed from the GUI thread plus one reader thread per peer connection.
        self._lock = threading.Lock()
        self.conn = sqlite3.connect(self.db_path, check_same_thread=False)
        self.conn.execute("PRAGMA journal_mode=WAL")
        self._init_schema()

    def _init_schema(self) -> None:
        self.conn.executescript(
            """
            CREATE TABLE IF NOT EXISTS contacts (
                fingerprint TEXT PRIMARY KEY,
                username TEXT NOT NULL,
                identity_pub BLOB NOT NULL,
                host TEXT,
                port INTEGER,
                verified INTEGER NOT NULL DEFAULT 0,
                added_at REAL NOT NULL
            );
            CREATE TABLE IF NOT EXISTS sessions (
                fingerprint TEXT PRIMARY KEY,
                encrypted_state BLOB NOT NULL,
                updated_at REAL NOT NULL,
                FOREIGN KEY(fingerprint) REFERENCES contacts(fingerprint)
            );
            CREATE TABLE IF NOT EXISTS messages (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                fingerprint TEXT NOT NULL,
                direction TEXT NOT NULL CHECK(direction IN ('in','out')),
                kind TEXT NOT NULL DEFAULT 'text',
                encrypted_body BLOB NOT NULL,
                timestamp REAL NOT NULL,
                FOREIGN KEY(fingerprint) REFERENCES contacts(fingerprint)
            );
            CREATE TABLE IF NOT EXISTS groups (
                group_id TEXT PRIMARY KEY,
                name TEXT NOT NULL,
                encrypted_state BLOB NOT NULL,
                updated_at REAL NOT NULL
            );
            CREATE TABLE IF NOT EXISTS group_messages (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                group_id TEXT NOT NULL,
                sender_identity_pub TEXT NOT NULL,
                kind TEXT NOT NULL DEFAULT 'text',
                encrypted_body BLOB NOT NULL,
                timestamp REAL NOT NULL,
                FOREIGN KEY(group_id) REFERENCES groups(group_id)
            );
            """
        )
        self.conn.commit()

    # -- contacts ----------------------------------------------------

    def upsert_contact(
        self, fingerprint: str, username: str, identity_pub: bytes, host: str, port: int
    ) -> None:
        with self._lock:
            self.conn.execute(
                """
                INSERT INTO contacts (fingerprint, username, identity_pub, host, port, verified, added_at)
                VALUES (?, ?, ?, ?, ?, 0, ?)
                ON CONFLICT(fingerprint) DO UPDATE SET
                    username=excluded.username,
                    host=CASE WHEN excluded.host != '' THEN excluded.host ELSE contacts.host END,
                    port=CASE WHEN excluded.port != 0 THEN excluded.port ELSE contacts.port END
                """,
                (fingerprint, username, identity_pub, host, port, time.time()),
            )
            self.conn.commit()

    def set_verified(self, fingerprint: str, verified: bool = True) -> None:
        with self._lock:
            self.conn.execute(
                "UPDATE contacts SET verified=? WHERE fingerprint=?", (int(verified), fingerprint)
            )
            self.conn.commit()

    def get_contact(self, fingerprint: str) -> sqlite3.Row | None:
        with self._lock:
            self.conn.row_factory = sqlite3.Row
            cur = self.conn.execute("SELECT * FROM contacts WHERE fingerprint=?", (fingerprint,))
            return cur.fetchone()

    def list_contacts(self) -> list[sqlite3.Row]:
        with self._lock:
            self.conn.row_factory = sqlite3.Row
            cur = self.conn.execute("SELECT * FROM contacts ORDER BY username")
            return cur.fetchall()

    # -- ratchet session state ----------------------------------------

    def save_session(self, fingerprint: str, session: crypto.RatchetSession) -> None:
        state = json.dumps(
            {
                "send_chain_key": session.send_chain_key.hex(),
                "recv_chain_key": session.recv_chain_key.hex(),
                "send_index": session.send_index,
                "recv_index": session.recv_index,
            }
        ).encode("utf-8")
        blob = crypto.encrypt_authenticated(self.storage_key, state, aad=fingerprint.encode())
        with self._lock:
            self.conn.execute(
                """
                INSERT INTO sessions (fingerprint, encrypted_state, updated_at) VALUES (?, ?, ?)
                ON CONFLICT(fingerprint) DO UPDATE SET encrypted_state=excluded.encrypted_state,
                    updated_at=excluded.updated_at
                """,
                (fingerprint, blob, time.time()),
            )
            self.conn.commit()

    def load_session(self, fingerprint: str) -> crypto.RatchetSession | None:
        with self._lock:
            cur = self.conn.execute(
                "SELECT encrypted_state FROM sessions WHERE fingerprint=?", (fingerprint,)
            )
            row = cur.fetchone()
        if row is None:
            return None
        state = json.loads(
            crypto.decrypt_authenticated(self.storage_key, row[0], aad=fingerprint.encode())
        )
        return crypto.RatchetSession(
            send_chain_key=bytes.fromhex(state["send_chain_key"]),
            recv_chain_key=bytes.fromhex(state["recv_chain_key"]),
            send_index=state["send_index"],
            recv_index=state["recv_index"],
        )

    # -- messages -------------------------------------------------------

    def save_message(
        self, fingerprint: str, direction: str, plaintext: str, kind: str = "text"
    ) -> None:
        body = crypto.encrypt_authenticated(
            self.storage_key, plaintext.encode("utf-8"), aad=fingerprint.encode()
        )
        with self._lock:
            self.conn.execute(
                "INSERT INTO messages (fingerprint, direction, kind, encrypted_body, timestamp) "
                "VALUES (?, ?, ?, ?, ?)",
                (fingerprint, direction, kind, body, time.time()),
            )
            self.conn.commit()

    def history(self, fingerprint: str, limit: int = 500) -> list[dict]:
        with self._lock:
            cur = self.conn.execute(
                "SELECT direction, kind, encrypted_body, timestamp FROM messages "
                "WHERE fingerprint=? ORDER BY id ASC LIMIT ?",
                (fingerprint, limit),
            )
            rows = cur.fetchall()
        out = []
        for direction, kind, body, ts in rows:
            plaintext = crypto.decrypt_authenticated(
                self.storage_key, body, aad=fingerprint.encode()
            ).decode("utf-8")
            out.append({"direction": direction, "kind": kind, "text": plaintext, "ts": ts})
        return out

    # -- groups (Phase 3) -------------------------------------------------

    def save_group(self, group_id: str, name: str, state: dict) -> None:
        """`state` holds everything sensitive about the group (member list,
        my outgoing sender-key chain, every other member's incoming chain)
        as one JSON blob, encrypted the same way ratchet session state is —
        with the identity-derived local storage key, never the password."""
        blob = crypto.encrypt_authenticated(
            self.storage_key, json.dumps(state).encode("utf-8"), aad=group_id.encode()
        )
        with self._lock:
            self.conn.execute(
                """
                INSERT INTO groups (group_id, name, encrypted_state, updated_at) VALUES (?, ?, ?, ?)
                ON CONFLICT(group_id) DO UPDATE SET
                    name=excluded.name, encrypted_state=excluded.encrypted_state, updated_at=excluded.updated_at
                """,
                (group_id, name, blob, time.time()),
            )
            self.conn.commit()

    def load_group(self, group_id: str) -> dict | None:
        with self._lock:
            row = self.conn.execute(
                "SELECT encrypted_state FROM groups WHERE group_id=?", (group_id,)
            ).fetchone()
        if row is None:
            return None
        return json.loads(crypto.decrypt_authenticated(self.storage_key, row[0], aad=group_id.encode()))

    def list_groups(self) -> list[tuple[str, str]]:
        with self._lock:
            cur = self.conn.execute("SELECT group_id, name FROM groups ORDER BY name")
            return cur.fetchall()

    def save_group_message(
        self, group_id: str, sender_identity_pub_hex: str, plaintext: str, kind: str = "text"
    ) -> None:
        body = crypto.encrypt_authenticated(
            self.storage_key, plaintext.encode("utf-8"), aad=group_id.encode()
        )
        with self._lock:
            self.conn.execute(
                "INSERT INTO group_messages (group_id, sender_identity_pub, kind, encrypted_body, timestamp) "
                "VALUES (?, ?, ?, ?, ?)",
                (group_id, sender_identity_pub_hex, kind, body, time.time()),
            )
            self.conn.commit()

    def group_history(self, group_id: str, limit: int = 500) -> list[dict]:
        with self._lock:
            cur = self.conn.execute(
                "SELECT sender_identity_pub, kind, encrypted_body, timestamp FROM group_messages "
                "WHERE group_id=? ORDER BY id ASC LIMIT ?",
                (group_id, limit),
            )
            rows = cur.fetchall()
        out = []
        for sender_pub_hex, kind, body, ts in rows:
            plaintext = crypto.decrypt_authenticated(
                self.storage_key, body, aad=group_id.encode()
            ).decode("utf-8")
            out.append({"sender_identity_pub": sender_pub_hex, "kind": kind, "text": plaintext, "ts": ts})
        return out

    def close(self) -> None:
        with self._lock:
            self.conn.close()

File: webapp/static/js/storage.js
/**
 * Browser-side encrypted local storage — the IndexedDB equivalent of
 * haven/storage.py. Same design: message content is decrypted once on
 * receipt and re-encrypted at rest with a key derived from the identity
 * key (never the password), independent of the wire ratchet.
 */

const HavenStorage = (() => {
  "use strict";
  const H = Haven;

  function openDb(username) {
    return new Promise((resolve, reject) => {
      const req = indexedDB.open(`haven-${username}`, 2);
      req.onupgradeneeded = () => {
        const db = req.result;
        if (!db.objectStoreNames.contains("contacts")) {
          db.createObjectStore("contacts", { keyPath: "fingerprint" });
        }
        if (!db.objectStoreNames.contains("sessions")) {
          db.createObjectStore("sessions", { keyPath: "fingerprint" });
        }
        if (!db.objectStoreNames.contains("messages")) {
          const store = db.createObjectStore("messages", { keyPath: "id", autoIncrement: true });
          store.createIndex("fingerprint", "fingerprint", { unique: false });
        }
        if (!db.objectStoreNames.contains("groups")) {
          db.createObjectStore("groups", { keyPath: "groupId" });
        }
        if (!db.objectStoreNames.contains("groupMessages")) {
          const store = db.createObjectStore("groupMessages", { keyPath: "id", autoIncrement: true });
          store.createIndex("groupId", "groupId", { unique: false });
        }
      };
      req.onsuccess = () => resolve(req.result);
      req.onerror = () => reject(req.error);
    });
  }

  function tx(db, storeName, mode, fn) {
    return new Promise((resolve, reject) => {
      const t = db.transaction(storeName, mode);
      const store = t.objectStore(storeName);
      const result = fn(store);
      t.oncomplete = () => resolve(result);
      t.onerror = () => reject(t.error);
    });
  }

  function reqToPromise(req) {
    return new Promise((resolve, reject) => {
      req.onsuccess = () => resolve(req.result);
      req.onerror = () => reject(req.error);
    });
  }

  class Store {
    constructor(db, storageKey) {
      this.db = db;
      this.storageKey = storageKey;
    }

    static async open(username, identityPrivateBytes) {
      const db = await openDb(username);
      const storageKey = await H.hkdf(identityPrivateBytes, H.utf8("local-storage-v1"));
      return new Store(db, storageKey);
    }

    // `verified` defaults to preserving whatever was already stored
    // rather than to false — this is called on every relay
    // (re)connection (see network.js's _registerConnection), and a
    // literal `= false` default here would silently wipe out a
    // contact's safety-number verification on every reconnect.
    async upsertContact(fingerprint, username, identityPubHex, verified) {
      const existing = await this.getContact(fingerprint);
      const record = {
        fingerprint,
        username,
        identityPubHex,
        verified: verified !== undefined ? verified : existing ? existing.verified : false,
        avatarDataUrl: existing ? existing.avatarDataUrl : undefined,
        addedAt: existing ? existing.addedAt : Date.now(),
      };
      await tx(this.db, "contacts", "readwrite", (store) => store.put(record));
    }

    async setVerified(fingerprint, verified) {
      const existing = await this.getContact(fingerprint);
      if (!existing) return;
      existing.verified = verified;
      await tx(this.db, "contacts", "readwrite", (store) => store.put(existing));
    }

    async setAvatar(fingerprint, avatarDataUrl) {
      const existing = await this.getContact(fingerprint);
      if (!existing) return;
      existing.avatarDataUrl = avatarDataUrl;
      await tx(this.db, "contacts", "readwrite", (store) => store.put(existing));
    }

    async getContact(fingerprint) {
      const t = this.db.transaction("contacts", "readonly");
      return reqToPromise(t.objectStore("contacts").get(fingerprint));
    }

    async listContacts() {
      const t = this.db.transaction("contacts", "readonly");
      return reqToPromise(t.objectStore("contacts").getAll());
    }

    async saveSession(fingerprint, session) {
      const state = {
        sendChainKeyHex: H.bytesToHex(session.sendChainKey),
        recvChainKeyHex: H.bytesToHex(session.recvChainKey),
        sendIndex: session.sendIndex,
        recvIndex: session.recvIndex,
      };
      const blob = await H.encryptAuthenticated(this.storageKey, H.utf8(JSON.stringify(state)), H.utf8(fingerprint));
      await tx(this.db, "sessions", "readwrite", (store) => store.put({ fingerprint, blobHex: H.bytesToHex(blob) }));
    }

    async loadSession(fingerprint) {
      const t = this.db.transaction("sessions", "readonly");
      const row = await reqToPromise(t.objectStore("sessions").get(fingerprint));
      if (!row) return null;
      const plaintext = await H.decryptAuthenticated(this.storageKey, H.hexToBytes(row.blobHex), H.utf8(fingerprint));
      const state = JSON.parse(H.fromUtf8(plaintext));
      return new H.RatchetSession(
        H.hexToBytes(state.sendChainKeyHex),
        H.hexToBytes(state.recvChainKeyHex),
        state.sendIndex,
        state.recvIndex
      );
    }

    // Returns the row's local auto-increment id — see network.js's
    // sendText/_handleMsg, which thread it through to the UI so a
    // rendered bubble can be found again later (to update its seen-tick,
    // or to pin/reply-target it). msgIndex is the ratchet envelope index
    // this plaintext was sent/received at (only meaningful for "out"
    // messages here — see markSeenUpTo, which uses it as a "read up to"
    // watermark instead of needing a per-message read-receipt id).
    async saveMessage(fingerprint, direction, plaintext, kind = "text", timestamp = Date.now(), msgIndex = null) {
      const blob = await H.encryptAuthenticated(this.storageKey, H.utf8(plaintext), H.utf8(fingerprint));
      const req = await tx(this.db, "messages", "readwrite", (store) =>
        store.add({ fingerprint, direction, kind, blobHex: H.bytesToHex(blob), timestamp, msgIndex, seen: false, pinned: false })
      );
      return req.result;
    }

    async history(fingerprint) {
      const t = this.db.transaction("messages", "readonly");
      const index = t.objectStore("messages").index("fingerprint");
      const rows = await reqToPromise(index.getAll(fingerprint));
      const out = [];
      for (const row of rows) {
        const plaintext = await H.decryptAuthenticated(this.storageKey, H.hexToBytes(row.blobHex), H.utf8(fingerprint));
        out.push({
          id: row.id,
          direction: row.direction,
          kind: row.kind,
          text: H.fromUtf8(plaintext),
          ts: row.timestamp,
          seen: !!row.seen,
          pinned: !!row.pinned,
        });
      }
      return out;
    }

    // Marks every outgoing message with a ratchet index <= upToIndex as
    // seen — the recipient sends this "read up to" watermark (their
    // session's recvIndex, see network.js) rather than acking each
    // message individually, which is simpler and self-healing (one lost
    // receipt doesn't leave a message stuck "unseen" forever, the next
    // receipt covers it too).
    async markSeenUpTo(fingerprint, upToIndex) {
      return new Promise((resolve, reject) => {
        const t = this.db.transaction("messages", "readwrite");
        const req = t.objectStore("messages").index("fingerprint").openCursor(IDBKeyRange.only(fingerprint));
        req.onsuccess = () => {
          const cursor = req.result;
          if (!cursor) return;
          const row = cursor.value;
          if (row.direction === "out" && row.msgIndex != null && row.msgIndex <= upToIndex && !row.seen) {
            row.seen = true;
            cursor.update(row);
          }
          cursor.continue();
        };
        req.onerror = () => reject(req.error);
        t.oncomplete = () => resolve();
        t.onerror = () => reject(t.error);
      });
    }

    async setPinned(id, pinned) {
      return new Promise((resolve, reject) => {
        const t = this.db.transaction("messages", "readwrite");
        const store = t.objectStore("messages");
        const getReq = store.get(id);
        getReq.onsuccess = () => {
          const row = getReq.result;
          if (!row) return;
          row.pinned = pinned;
          store.put(row);
        };
        getReq.onerror = () => reject(getReq.error);
        t.oncomplete = () => resolve();
        t.onerror = () => reject(t.error);
      });
    }

    // "Clear chat" — wipes this contact's message history but keeps the
    // contact and its session, so the chat is simply empty afterward
    // rather than gone. See deleteContact for the more destructive
    // "delete chat" operation.
    async clearMessages(fingerprint) {
      return new Promise((resolve, reject) => {
        const t = this.db.transaction("messages", "readwrite");
        const req = t.objectStore("messages").index("fingerprint").openCursor(IDBKeyRange.only(fingerprint));
        req.onsuccess = () => {
          const cursor = req.result;
          if (!cursor) return;
          cursor.delete();
          cursor.continue();
        };
        req.onerror = () => reject(req.error);
        t.oncomplete = () => resolve();
        t.onerror = () => reject(t.error);
      });
    }

    // "Delete chat" — this is a local, per-device operation (same as
    // every other privacy setting here): it forgets the contact and its
    // session on this device only, it does not notify them or affect
    // their copy of the conversation. If they message you again, a new
    // session simply gets re-established and they reappear as a contact.
    async deleteContact(fingerprint) {
      await this.clearMessages(fingerprint);
      await tx(this.db, "contacts", "readwrite", (store) => store.delete(fingerprint));
      await tx(this.db, "sessions", "readwrite", (store) => store.delete(fingerprint));
    }

    // Every message across every contact, decrypted — used only for
    // building a full-account backup (see backup.js), not the normal
    // per-chat history() path above.
    async listAllMessages() {
      const t = this.db.transaction("messages", "readonly");
      const rows = await reqToPromise(t.objectStore("messages").getAll());
      const out = [];
      for (const row of rows) {
        const plaintext = await H.decryptAuthenticated(this.storageKey, H.hexToBytes(row.blobHex), H.utf8(row.fingerprint));
        out.push({ fingerprint: row.fingerprint, direction: row.direction, kind: row.kind, text: H.fromUtf8(plaintext), ts: row.timestamp });
      }
      return out;
    }

    // Group state (membership + sender-key chains) is encrypted the same
    // way a 1:1 session is — see groups.js for what "state" contains.
    async saveGroup(groupId, name, state) {
      const blob = await H.encryptAuthenticated(this.storageKey, H.utf8(JSON.stringify(state)), H.utf8(groupId));
      await tx(this.db, "groups", "readwrite", (store) => {
        store.put({ groupId, name, blobHex: H.bytesToHex(blob), updatedAt: Date.now() });
      });
    }

    async loadGroup(groupId) {
      const t = this.db.transaction("groups", "readonly");
      const row = await reqToPromise(t.objectStore("groups").get(groupId));
      if (!row) return null;
      const plaintext = await H.decryptAuthenticated(this.storageKey, H.hexToBytes(row.blobHex), H.utf8(groupId));
      return { name: row.name, state: JSON.parse(H.fromUtf8(plaintext)) };
    }

    async listGroups() {
      const t = this.db.transaction("groups", "readonly");
      const rows = await reqToPromise(t.objectStore("groups").getAll());
      return rows.map((r) => ({ groupId: r.groupId, name: r.name }));
    }

    async saveGroupMessage(groupId, senderIdentityPubHex, plaintext, kind = "text", timestamp = Date.now()) {
      const blob = await H.encryptAuthenticated(this.storageKey, H.utf8(plaintext), H.utf8(groupId));
      const req = await tx(this.db, "groupMessages", "readwrite", (store) =>
        store.add({ groupId, senderIdentityPubHex, kind, blobHex: H.bytesToHex(blob), timestamp, pinned: false })
      );
      return req.result;
    }

    async groupHistory(groupId) {
      const t = this.db.transaction("groupMessages", "readonly");
      const index = t.objectStore("groupMessages").index("groupId");
      const rows = await reqToPromise(index.getAll(groupId));
      const out = [];
      for (const row of rows) {
        const plaintext = await H.decryptAuthenticated(this.storageKey, H.hexToBytes(row.blobHex), H.utf8(groupId));
        out.push({
          id: row.id,
          senderIdentityPubHex: row.senderIdentityPubHex,
          kind: row.kind,
          text: H.fromUtf8(plaintext),
          ts: row.timestamp,
          pinned: !!row.pinned,
        });
      }
      return out;
    }

    async setGroupMessagePinned(id, pinned) {
      return new Promise((resolve, reject) => {
        const t = this.db.transaction("groupMessages", "readwrite");
        const store = t.objectStore("groupMessages");
        const getReq = store.get(id);
        getReq.onsuccess = () => {
          const row = getReq.result;
          if (!row) return;
          row.pinned = pinned;
          store.put(row);
        };
        getReq.onerror = () => reject(getReq.error);
        t.oncomplete = () => resolve();
        t.onerror = () => reject(t.error);
      });
    }

    async clearGroupMessages(groupId) {
      return new Promise((resolve, reject) => {
        const t = this.db.transaction("groupMessages", "readwrite");
        const req = t.objectStore("groupMessages").index("groupId").openCursor(IDBKeyRange.only(groupId));
        req.onsuccess = () => {
          const cursor = req.result;
          if (!cursor) return;
          cursor.delete();
          cursor.continue();
        };
        req.onerror = () => reject(req.error);
        t.oncomplete = () => resolve();
        t.onerror = () => reject(t.error);
      });
    }

    // Local-only, same as deleteContact: forgets this group's messages
    // and membership state on this device without notifying anyone.
    async deleteGroup(groupId) {
      await this.clearGroupMessages(groupId);
      await tx(this.db, "groups", "readwrite", (store) => store.delete(groupId));
    }
  }

  return { Store };
})();
